core/src/main/scala-3.0.0-RC3/cats/derived/hash.scala

```diff
@@ -0,0 +1,36 @@
+package cats.derived
+
+import cats.Hash
+import shapeless3.deriving.{K0, Continue}
+
+import scala.util.hashing.MurmurHash3
+
+object hash extends HashDerivation
+
+trait ProductHash[T[x] <: Hash[x], A](using inst: K0.ProductInstances[T, A])
+  extends ProductEq[T, A], Hash[A]:
+
+  def hash(x: A): Int = {
+    val (hash, len) = inst.foldLeft[(Int, Int)](x)((MurmurHash3.productSeed, 0))(
+      [t] => (acc: (Int, Int), h: T[t], t0: t) => Continue((MurmurHash3.mix(acc._1, h.hash(t0)), acc._2 + 1))
+    )
```

Comment:
I would rather do two iterations without allocating tuples

Reply:
Oh, even better: restrict `A <: Product` and use `productArity` directly

Reply:
Oh good call! 👍

Reply:
Hmm I gather that `A <: Product` doesn't quite work so we need `A <:< Product`.
I wonder if we can avoid the type casts somehow.

Reply:
I wouldn't want to guarantee that we can't 😂 But I couldn't see how to get `derived` to work, where we need `A <: Product` only if `Mirror[A]` is a `Product` and not a `Sum`